runtime/src/main/java/org/corfudb/runtime/collections/CorfuTable.java

```diff
@@ -228,13 +227,13 @@ public V get(@ConflictParameter Object key) {
      *
      * @param indexName Name of the the secondary index to query.
      * @param indexKey  The index key used to query the secondary index
-     * @return A collection of Map.Entry<K, V>
+     * @return A stream of {@code Map.Entry<K, V>}
      */
     @SuppressWarnings("unchecked")
     @Accessor
     public @Nonnull
     <I>
-    Collection<Entry<K, V>> getByIndex(@Nonnull Index.Name indexName, I indexKey) {
+    Stream<Entry<K, V>> getByIndex(@Nonnull Index.Name indexName, I indexKey) {
         String secondaryIndex = indexName.get();
         Map<Object, Map<K, V>> secondaryMap;
```

Comment:
What is the actual implementation for `Map<K, V`? `HashMap`?

Reply:
Yes, it's a `HashMap`.

Reply:
If it's a `HashMap`, then there are two scenarios:

1. The clients wants a collection.
2. The clients wants a stream.

In case of (1), you'll end up converting from a `HashSet` to a `Stream` and back to a `HashSet`, which seems less efficient than the existing approach. Are we trying to reduce boiler-plate code, or improve performance?

Reply:
The goal of this patch is to reduce the amount of temporary objects created. In all current use cases, the collection produced by this API gets turned into a stream, then back into a collection after filtering/mapping operations are performed. If any other collection is desired, one still needs to convert from HashSet to said collection.

Reply:
Even in the case where a HashSet is the desired end result (with no additional filtering or mapping), the current implementation will call addAll and iterate through the entrySet to populate this new HashSet.